### PR TITLE
plugins/telekasten: init

### PR DESCRIPTION
**Pull request by onemoresuza**
Add [telekasten](https://github.com/nvim-telekasten/telekasten.nvim), a
Zettelkasten implementation using telescope.

### CODE REVIEW

File: tests/test-sources/plugins/by-name/telekasten/default.nix
Comment (khaneliman) [inline tests/test-sources/plugins/by-name/telekasten/default.nix:3]:
```suggestion
    plugins.telekasten = {
```

File: tests/test-sources/plugins/by-name/telekasten/default.nix
Comment (khaneliman) [inline tests/test-sources/plugins/by-name/telekasten/default.nix:1]:
```suggestion
{
  empty = {
    plugins.telekasten.enable = true;
  };

```

File: tests/test-sources/plugins/by-name/telekasten/default.nix
Comment (onemoresuza) [reply tests/test-sources/plugins/by-name/telekasten/default.nix]: Oops, was thinking of different plugin. It's now fixed.

File: tests/test-sources/plugins/by-name/telekasten/default.nix
Comment (onemoresuza) [reply tests/test-sources/plugins/by-name/telekasten/default.nix:1]: Done.

**Review by MattSturgeon — DISMISSED** [review]
Awesome, thanks! :tada: 

<details><summary>Draft announcement</summary>
<p>


```md
\[💾 NEW PLUGIN\]

[telekasten.nvim](https://github.com/nvim-telekasten/telekasten.nvim) support has been added !

- **Description:** A Neovim (lua) plugin for working with a markdown zettelkasten / wiki and mixing it with a journal, based on telescope.nvim 
- [Documentation](https://nix-community.github.io/nixvim/plugins/telekasten.html)
- [PR](https://github.com/nix-community/nixvim/pull/2429) by [onemoresuza](https://github.com/onemoresuza)
```

</p>
</details>

**Comment by MattSturgeon** [comment]
@mergify queue

**Comment by MattSturgeon** [comment]
> E5113: Error while calling lua chunk: ...yNeovimPackages/start/telekasten.nvim/lua/telekasten.lua:1: module 'telescope.builtin' not found

https://buildbot.nix-community.org/#/builders/2871/builds/1421/steps/1/logs/stdio

Looks like there's a dependency on telescope, we'll need to handle that. @khaneliman should this be done with warnings/assertions?

**Comment by khaneliman** [comment]
> > E5113: Error while calling lua chunk: ...yNeovimPackages/start/telekasten.nvim/lua/telekasten.lua:1: module 'telescope.builtin' not found
> 
> https://buildbot.nix-community.org/#/builders/2871/builds/1421/steps/1/logs/stdio
> 
> Looks like there's a dependency on telescope, we'll need to handle that. @khaneliman should this be done with warnings/assertions?

If it's required, i'd do an assertion.

**Comment by onemoresuza** [comment]
I've checked the [overrides.nix](https://github.com/NixOS/nixpkgs/blob/73bd40281687e5bd0dadb63b34865231fd20e7e4/pkgs/applications/editors/vim/plugins/overrides.nix) file on nixpkgs and saw no mention of telekasten.

The package on nixpkgs should already add telescope as a dependecy --- [here's an example](https://github.com/NixOS/nixpkgs/blob/73bd40281687e5bd0dadb63b34865231fd20e7e4/pkgs/applications/editors/vim/plugins/overrides.nix#L146).

I'm going to make a PR to nixpkgs to fix it.

Should we wait for it to be merged and make this a draft or could the `extraPackages` attribute be used to circumvent this issue?

**Comment by khaneliman** [comment]
> I've checked the [overrides.nix](https://github.com/NixOS/nixpkgs/blob/73bd40281687e5bd0dadb63b34865231fd20e7e4/pkgs/applications/editors/vim/plugins/overrides.nix) file on nixpkgs and saw no mention of telekasten.
> 
> The package on nixpkgs should already add telescope as a dependecy --- [here's an example](https://github.com/NixOS/nixpkgs/blob/73bd40281687e5bd0dadb63b34865231fd20e7e4/pkgs/applications/editors/vim/plugins/overrides.nix#L146).
> 
> I'm going to make a PR to nixpkgs to fix it.
> 
> Should we wait for it to be merged and make this a draft or could the `extraPackages` attribute be used to circumvent this issue?

Yeah, it's best to upstream a hard dependency like that. Thank you.

**Comment by khaneliman** [comment]
> I'm going to make a PR to nixpkgs to fix it and, for this PR, add the assertion propsed by @khaneliman.

Sounds good to me.

**Review by MattSturgeon — APPROVED** [review]

**Review by khaneliman — APPROVED** [review]
LGTM thanks for your contribution!

**Comment by onemoresuza** [comment]
I see that there's new error now on the build logs: The assertion for the telescope plugin. 

Should I add `plugins.telescope.enable = true;` to either the example or the empty attributes of the plugin test?

**Comment by khaneliman** [comment]
> I see that there's new error now on the build logs: The assertion for the telescope plugin.
> 
> Should I add `plugins.telescope.enable = true;` to either the example or the empty attributes of the plugin test?

Lol, yes... sorry, forgot to say we needed to add setting the plugin to those tests.

**Comment by onemoresuza** [comment]
While creating the nixpkgs PR, I've also discovered that the plugin relies on plenary-nvim: Going to add it to extraPlugins.

**Comment by MattSturgeon** [comment]
> I've also discovered that the plugin relies on plenary-nvim: Going to add it to extraPlugins.

We prefer to have nullable package options to allow users to disable automatic installation; maybe they wish to handle the dep in a different way?

We would usually have an `extraOptions.plenaryPackage` option, that is a nullable `lib.mkPackageOption` style option and then add `cfg.plenaryPackage` to `extraPlugins` instead.

I suspect there's an example of this strategy somewhere in nixvim :slightly_smiling_face:

**Comment by onemoresuza** [comment]
> > I've also discovered that the plugin relies on plenary-nvim: Going to add it to extraPlugins.
> 
> We prefer to have nullable package options to allow users to disable automatic installation; maybe they wish to handle the dep in a different way?
> 
> We would usually have an `extraOptions.plenaryPackage` option, that is a nullable `lib.mkPackageOption` style option and then add `cfg.plenaryPackage` to `extraPlugins` instead.
> 
> I suspect there's an example of this strategy somewhere in nixvim 🙂

I see. I've found an example [here](https://github.com/nix-community/nixvim/blob/717e7060fafa2c3822a64e3f5bbfd4895577fdbf/plugins/lsp/language-servers/default.nix#L18). Going to update the PR.

**Comment by onemoresuza** [comment]
A new error appeared on the [logs](https://buildbot.nix-community.org/#/builders/2807/builds/1437) related to `plugins.web-devicons` being enabled automatically.

I've found [this](https://github.com/nix-community/nixvim/blob/717e7060fafa2c3822a64e3f5bbfd4895577fdbf/plugins/by-name/telescope/default.nix#L99) on the telescope implementation.

Should I enable `plugins.web-devicons` for now, and the remove on the 24.11 release?

**Comment by onemoresuza** [comment]
Here's the upstream nixpkgs PR: https://github.com/NixOS/nixpkgs/pull/349346#issue-2595535643

**Comment by MattSturgeon** [comment]
> A new error appeared on the [logs](https://buildbot.nix-community.org/#/builders/2807/builds/1437) related to `plugins.web-devicons` being enabled automatically.
> 
> I've found [this](https://github.com/nix-community/nixvim/blob/717e7060fafa2c3822a64e3f5bbfd4895577fdbf/plugins/by-name/telescope/default.nix#L99) on the telescope implementation.
> 
> Should I enable `plugins.web-devicons` for now, and the remove on the 24.11 release?

We do want to silence the warning, yes. I believe explicitly _disabling_ web-devicons would also silence it? i.e. any definition, true or false.



> Here's the upstream nixpkgs PR: [NixOS/nixpkgs#349346 (comment)](https://github.com/NixOS/nixpkgs/pull/349346#issue-2595535643)

Awesome, thanks! Do you want to reference that PR in TODO comments for anything we could remove once it is in nixos-unstable?

**Comment by onemoresuza** [comment]
> We do want to silence the warning, yes. I believe explicitly _disabling_ web-devicons would also silence it? i.e. any definition, true or false.

Going to explicitly disable it then.

> Awesome, thanks! Do you want to reference that PR in TODO comments for anything we could remove once it is in nixos-unstable?

Going to add TODOs regarding both plenary and telescope, since they will be handled by the upstream package once the PR lands on nixpkgs.

**Comment by onemoresuza** [comment]
Checks succeeded!

**Review by khaneliman — APPROVED** [review]
LGTM thanks for handling the upstream dependencies, too.

**Comment by khaneliman** [comment]
@mergify queue

**Pull request merged**
